util/cloud/s3.cc

```diff
@@ -95,33 +98,36 @@ std::pair<size_t, string_view> ParseXmlObjContents(xmlNodePtr node) {
   return res;
 }
 
-error_code ParseListObj(string_view xml_obj, S3Bucket::ListObjectCb cb) {
+ListObjectsResult ParseListObj(string_view xml_obj, S3Bucket::ListObjectCb cb) {
   xmlDocPtr doc = XmlRead(xml_obj);
 
+  bool truncated = false;
+  string_view last_key = "";
+
   if (!doc) {
     LOG(ERROR) << "Could not parse xml response " << xml_obj;
-    return make_error_code(errc::bad_message);
+    return make_unexpected(make_error_code(errc::bad_message));
   }
 
+  absl::Cleanup xml_free{[doc]() { xmlFreeDoc(doc); }};
+
   xmlNodePtr root = xmlDocGetRootElement(doc);
   CHECK_STREQ("ListBucketResult", as_char(root->name));
-
   for (xmlNodePtr child = root->children; child; child = child->next) {
     if (child->type == XML_ELEMENT_NODE) {
       xmlNodePtr grand = child->children;
       if (!strcmp(as_char(child->name), "IsTruncated")) {
-        CHECK(grand && grand->type == XML_TEXT_NODE);
-        CHECK_STREQ("false", as_char(grand->content)) << "TBD: need to support cursor based functionality";
+        truncated = strcmp(as_char(grand->content), "true") == 0;
       } else if (!strcmp(as_char(child->name), "Marker")) {
       } else if (!strcmp(as_char(child->name), "Contents")) {
```

Comment:
`== "const"sv` looks cleaner, `== 0` is easier to understand. But I'll change it back to `truncated = !strcmp`